[CDAP-16236] Reverts the change to remove check for macros before running preview

ajainarayanan opened the pull request:
JIRA: https://issues.cask.co/browse/CDAP-16236
Build: https://builds.cask.co/browse/CDAP-UDUT579-1
E2E tests: https://builds.cask.co/browse/IT-UPD2141-1

**Problem:** 
- We accidentally removed the check for macros before running preview. This shouldn't have been merged.
- This was not caught by e2e tests (even though we had one) because is disabled in e2e testing environment

**Fix:**
- Revert the change
- Make tests run locally but skip in distributed environments.
- _Note_: The majority of change in this PR is in test and not in actual code.

**TODO:**
- Make preview enabled in e2e environments

ajainarayanan committed:
[CDAP-16236] Reverts the change to remove check for macros before running preview

 - Fixes preview tests to be skipped on environments where preview is disabled
 - Fixes data-cy attribute of run button in preview to be consistent
 - Show runtime arguments if there are unfulfilled macros before running preview

ajainarayanan commented:
Local run of runtime arguments tests
<img width="1440" alt="Screen Shot 2020-04-19 at 9 48 56 AM" src="https://user-images.githubusercontent.com/1452845/79694070-02c2ab80-8223-11ea-8cb2-1f9a38db0413.png">

itsanudeep commented on cdap-ui/app/hydrator/controllers/create/toppanel-ctrl.js:
```diff
@@ -502,6 +511,11 @@ class HydratorPlusPlusTopPanelCtrl {
+    this.validToStartPreview = this.isValidToStartPreview();
+  }
+
+  isValidToStartPreview() {
```
nit: Can we name this something more appropriate? Having empty values is still valid to start a preview, it's just we don't want to 'Run' the preview right away without opening the modeless.

itsanudeep commented on cdap-ui/cypress/integration/pipeline.runtimeargs.spec.ts:
```diff
@@ -29,16 +33,23 @@ const PREVIEW_SUCCESS_BANNER_MSG =
 
-/* Disabling preview tests for now because our E2E tests don't have preview enabled
- describe('Creating pipeline with macros ', () => {
+let skipPreviewTests = false;
```
Can we add a TODO to clean this up when we have preview enabled in bamboo?

itsanudeep commented on cdap-ui/cypress/integration/pipeline.runtimeargs.spec.ts:
```diff
@@ -48,21 +59,22 @@ const SOURCE_PATH_VAL = 'file:/tmp/cdap-ui-integration-fixtures/airports.csv';
-
-  it('and providing wrong runtime arguments should fail the pipeline preview.', () => {
+    );
+    cy.wait(10000);
```
We were trying to avoid explicit wait statements like these, right? You can use `cy.get('[title="Airport_source"').should('exist');` to check if the pipeline is deployed like we do in other tests?

ajainarayanan replied:
I agree we need to avoid the wait times but this is uploading a json in pipeline studio and we don't have a fail-proof way to find if its complete.

itsanudeep replied:
Won't `cy.get('[title="Airport_source"').should('exist'); ` do this? We're testing if title in source node is present, which we only see after the pipeline is uploaded and ready in studio?

itsanudeep reviewed (commented):
Can you update the PR description and remove regression label from the jira and comment the changes we made here? 

I do not think this is a regression. Previously, in preview, if the runtimes arguments are selected to be provided and the user directly clicks on run button, we start the preview without opening the modeless. 

Now that we have removed provided arguments checkbox and we assume if runtime arguments’ value is empty then those arguments would be provided. So, because of this we assume empty values to be valid and we start the pipeline run without opening the modeless.

I like opening the modeless before we start preview for arguments with empty values but I wouldn't call the previous behavior a regression. Please let me know if you feel otherwise.

itsanudeep commented:
minor comments, otherwise LGTM 👍

ajainarayanan commented:
> Can you update the PR description and remove regression label from the jira and comment the changes we made here?
> 
> I do not think this is a regression. Previously, in preview, if the runtimes arguments are selected to be provided and the user directly clicks on run button, we start the preview without opening the modeless.
> 
> Now that we have removed provided arguments checkbox and we assume if runtime arguments’ value is empty then those arguments would be provided. So, because of this we assume empty values to be valid and we start the pipeline run without opening the modeless.
> 
> I like opening the modeless before we start preview for arguments with empty values but I wouldn't call the previous behavior a regression. Please let me know if you feel otherwise.

@itsanudeep I don't think so. The reason why I tagged this as regression is,
1. Previously we used to show the runtime arguments modeless all the time if there are macros.
2. We right now always show the runtime arguments modeless all the time in pipeline detailed view
3. When we removed provided we just removed the requirement for filling up macros before running the pipeline. We implemented that correctly in deployed pipeline by showing the modeless everytime when the user starts a pipeline, however we didn't do the same in pipeline preview
4. So the exact regression is, someone running a preview before 6.1 we always show the macros and they can consciously show them the macros and let them run the preview and it is always predictable. But now we implicitly start the preview without opening the modeless and the preview fails and the user has no idea unless they open the configure modeless.

Preview runs fine before but now it always fails and then the user realized they need to provide macros that we didn't show them when running the preview.

itsanudeep commented:
> 1. Previously we used to show the runtime arguments modeless all the time if there are macros.
> 4. So the exact regression is, someone running a preview before 6.1 we always show the macros and they can consciously show them the macros and let them run the preview and it is always predictable. But now we implicitly start the preview without opening the modeless and the preview fails and the user has no idea unless they open the configure modeless.

I think you got this wrong. In preview, we do not open the modeless if arguments (macros) are selected as provided. Here is that behavior from before we made that change (6.1)

![image](https://user-images.githubusercontent.com/51677126/79904568-b5347300-83c9-11ea-9284-15db1d17b49d.png)

itsanudeep commented:
Oops, noticed previous gif was attached as an image. Here's a gif:

![not-regression](https://user-images.githubusercontent.com/51677126/79909752-5cb5a380-83d2-11ea-9f8f-c5146289ebe9.gif)

ajainarayanan commented:
> Oops, noticed previous gif was attached as an image. Here's a gif:
> 
> ![not-regression](https://user-images.githubusercontent.com/51677126/79909752-5cb5a380-83d2-11ea-9f8f-c5146289ebe9.gif)

@itsanudeep Hhhmm thats not true. The provided checkbox are NOT checked by default right? The user, when the run the preview, we open the modeless AND when they check the provided checkbox THEN we run the preview directly from subsequent runs. You still have to click the checkbox? Thats a conscious decision that the user action from the user. We went from this to directly running the preview. From here we regressed this behavior and started directly running the preview without letting the user know that there are macros that needs input.

And to worsen this, we shipped a patch release where we replaced the runtime arguments in pipeline detailed view and removed the checkbox but not entirely in preview. So in the exiting release we have preview runtime arguments where either the user can check ALL the runtime arguments or NONE. We didn't remove the header and we didn't remove the check. The right fix would have been to retain the checkbox for preview runtime arguments which was another regression.

itsanudeep reviewed (approved):
LGTM

ajainarayanan commented:
Green build: https://builds.cask.co/browse/CDAP-UDUT579-6
